pkg/ptrace/ptrace_linux.go

```diff
@@ -0,0 +1,418 @@
+	}
+
+	return errors.Errorf(waitPidErrorMessage, ret)
+}
+
+// Trace ptrace all threads of a process
+func Trace(pid int, logger logr.Logger) (*TracedProgram, error) {
+	traceSuccess := false
+
+	tidMap := make(map[int]bool)
+	retryCount := make(map[int]int)
+
+	// iterate over the thread group, until it doens't change
+	//
+	// we have tried several ways to ensure that we have stopped all the tasks:
+	// 1. iterating over and over again to make sure all of them are tracee
+	// 2. send `SIGSTOP` signal
+	// ...
+	// only the first way finally worked for every situations
+	for {
+		threads, err := ioutil.ReadDir(fmt.Sprintf("/proc/%d/task", pid))
+		if err != nil {
+			logger.Error(err, "read failed", "pid", pid)
+			return nil, errors.WithStack(err)
+		}
+
+		// judge whether `threads` is a subset of `tidMap`
+		subset := true
+
+		tids := make(map[int]bool)
+		for _, thread := range threads {
+			tid64, err := strconv.ParseInt(thread.Name(), 10, 32)
+			if err != nil {
+				return nil, errors.WithStack(err)
+			}
+			tid := int(tid64)
+
+			_, ok := tidMap[tid]
+			if ok {
+				tids[tid] = true
+				continue
+			}
+			subset = false
+
+			err = syscall.PtraceAttach(tid)
+			if err != nil {
+				_, ok := retryCount[tid]
+				if !ok {
+					retryCount[tid] = 1
+				} else {
+					retryCount[tid]++
+				}
+				if retryCount[tid] < threadRetryLimit {
+					logger.Info("retry attaching thread", "tid", tid, "retryCount", retryCount[tid], "limit", threadRetryLimit)
+					continue
+				}
+
+				if !strings.Contains(err.Error(), "no such process") {
+					logger.Error(err, "attach failed", "tid", tid)
+					return nil, errors.WithStack(err)
+				}
+				continue
+			}
+			defer func() {
+				if !traceSuccess {
+					err = syscall.PtraceDetach(tid)
+					if err != nil {
+						if !strings.Contains(err.Error(), "no such process") {
+							logger.Error(err, "detach failed", "tid", tid)
+						}
+					}
+				}
+			}()
+
+			err = waitPid(tid)
+			if err != nil {
+				return nil, errors.WithStack(err)
+			}
+
+			logger.Info("attach successfully", "tid", tid)
+			tids[tid] = true
+			tidMap[tid] = true
+		}
+
+		if subset {
+			tidMap = tids
+			break
+		}
+	}
+
+	var tids []int
+	for key := range tidMap {
+		tids = append(tids, key)
+	}
+
+	entries, err := mapreader.Read(pid)
+	if err != nil {
+		return nil, err
+	}
+
+	program := &TracedProgram{
+		pid:        pid,
+		tids:       tids,
+		Entries:    entries,
+		backupRegs: &syscall.PtraceRegs{},
+		backupCode: make([]byte, syscallInstrSize),
+		logger:     logger,
+	}
+
+	traceSuccess = true
+
+	return program, nil
+}
+
+// Detach detaches from all threads of the processes
+func (p *TracedProgram) Detach() error {
+	for _, tid := range p.tids {
+		p.logger.Info("detaching", "tid", tid)
+
+		err := syscall.PtraceDetach(tid)
+
+		if err != nil {
+			if !strings.Contains(err.Error(), "no such process") {
+				p.logger.Error(err, "detach failed", "tid", tid)
+				return errors.WithStack(err)
+			}
```

Comment:
Why print error log here?

Reply:
No. I shouldn't (they are careless copy-paste codes). 

I removed these log, and added a log error at the `time_server.go`.